### PR TITLE
Introduce Cloud docker variant

### LINKED ISSUE #74795: Build Elasticsearch Docker image with Cloud dependencies embedded
As discussed, we want to build an image that is directly deployable to Elastic Cloud. The intent is for this image to mostly identical to the existing Docker images with two main exceptions:

1. The `repository-azure`, `repository-gcs` and `repository-s3` plugins will be pre-installed.
2. Filebeat and Metricbeat will be installed under `/opt/filebeat` and `/opt/metricbeat`, respectively.

The first item should be pretty strait forward, we just want to install these plugins as part of building the image so they are already baked into the image for Cloud.

The second item will require slightly more coordination, we'll need to fetch these things from the Elastic snapshot artifacts repo when building locally (@breskeby can provide guidance here), and for release builds we'll need to use the release-manager built artifacts for those projects.

No need to update download pages or documentation here, this image isn't intended to be used by "normal" users. We also won't be publishing these to Dockerhub for the same reason.

### PR DESCRIPTION
**Pull request by pugnascotia**
Closes #74795.

Introduce two Docker image variants for Cloud. The first bundles (actually installs) the S3, Azure and GCS repository plugins. The second bundles all official plugins, but only installs the repository plugins.

Both images also bundle Filebeat and Metricbeat.

The testing utils have been refactored to introduce a `docker` sub-package. This allows the static `Docker.containerId` to be shared without needing all the code in one big class. The code for checking file ownership / permissions has also been refactored to a more Hamcrest style, using a custom Docker file matcher.

### CODE REVIEW

File: distribution/docker/src/docker/Dockerfile
Comment (mark-vieira) [inline distribution/docker/src/docker/Dockerfile:265]: Should we delete these files after install?

Comment (pugnascotia) [reply distribution/docker/src/docker/Dockerfile:265]: This happens in the `builder` image, which we throw away.

Comment (mark-vieira) [reply distribution/docker/src/docker/Dockerfile:265]: Ah, right.

File: distribution/docker/build.gradle
Comment (alpar-t) [inline distribution/docker/build.gradle:69]: Does this work itself out to a version with a build ID in case of release manager builds? I think we need that to guarantee consistency in the unified release process and make sure that we are picking up the beats from the same build. Also means that ES needs to be built after beats as part of that process. 

Also, someone on the beats team might want to customize this, e.g. check out some ES tag to get a stable/known to be good ES and build the docker images with a custom working version of beats. That's a use case for PRs against beats, but I think it's this code that needs to enable it somehow. Just a taught, it can be done in a follow up PR too.

File: distribution/docker/src/docker/Dockerfile
Comment (alpar-t) [inline distribution/docker/src/docker/Dockerfile:264]: This is a question for @mieciu ho's probably on top of this already. Will this work with the existing startup script in cloud ? Once we start using these images, they need to work with already released versions of ECE so while we can and should  make improvements to this script on the cloud side, we have to be backwards compatible too. I'm thinking we might also need a way to tell cloud here that these are already installed.

Comment (mieciu) [reply distribution/docker/src/docker/Dockerfile:264]: Indeed, that's something that's not going to work for us, at least for `7.x`.

Comment (pugnascotia) [reply distribution/docker/src/docker/Dockerfile:264]: Can you advise on what would work?

Comment (pugnascotia) [reply distribution/docker/src/docker/Dockerfile:264]: Perhaps `allocator.sh` could be patched to never add or remove these plugins?

Comment (alpar-t) [reply distribution/docker/src/docker/Dockerfile:264]: I think that's the way to go, I pinged @mieciu because I think that can be handled on the cloud side. We add this script as part of the stack-pack so we can make changes to it that will take affect on older allocators too.

File: distribution/docker/src/docker/Dockerfile
Comment (alpar-t) [inline distribution/docker/src/docker/Dockerfile:263]: Has there been a discussion around other plugins that are now part of the cloud image and customers can install ? 
AFAIK some of the popular examples are the `analysis-*` plugins.  We have options either way: bundle them in the image, but don't install them by default, or have cloud install them over the download service ( and document that it doesn't work for air gaped installation, these would install it like they install a custom plugin ). 7.x and ECE 2.x are think sensible to this approach we would need to seek PM approval as it would I think be a breaking change for such environments I'm not sure it's ok to make or not.

Comment (pugnascotia) [reply distribution/docker/src/docker/Dockerfile:263]: I wasn't involved in the discussions to date, I was going by #74795. Once we have something work end-to-end for these plugins, from the ES perspective it ought to be straightforward to add more.

**Comment by pugnascotia** [comment]
@elasticmachine update branch

File: distribution/docker/src/docker/Dockerfile
Comment (mark-vieira) [reply distribution/docker/src/docker/Dockerfile:263]: We discussed this an concluded that for non-essential plugins (ones that aren't _always_ installed for any given cloud provider) they would simply be downloaded as with an on-prem install.

File: distribution/docker/build.gradle
Comment (pugnascotia) [reply distribution/docker/build.gradle:77]: The corresponding release-manager changes fetch the beats artifacts using the build ID, so everything should match up.

File: distribution/docker/src/docker/Dockerfile
Comment (pugnascotia) [reply distribution/docker/src/docker/Dockerfile:265]: After further discussion, this PR now builds a `cloud` image and a `cloud-ess` image. The former image installs only the 3 repository plugins, and a plugin tool wrapper that checks in an archive directory for other plugins before reaching out to the internet. The latter image includes all the other plugins in the archive directory, in a "batteries-included" manner.

File: distribution/docker/build.gradle
Comment (mark-vieira) [inline distribution/docker/build.gradle:18]: We should use `providers.systemProperty()` here. Do a search to see examples of other usages in our build.

File: distribution/docker/build.gradle
Comment (mark-vieira) [inline distribution/docker/build.gradle:312]: Is it strictly necessary that we do this here? Isn't this done by release manager?

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DockerBase.java
Comment (mark-vieira) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DockerBase.java:27]: Do we actually use this base image bit anywhere. I feel like we end up overriding this in `docker/build.gradle` anyhow.

File: plugins/build.gradle
Comment (mark-vieira) [inline plugins/build.gradle:32]: This doesn't seem right. We're adding dependencies to these configurations here _and_ resolving them from another project, which I'm pretty sure is a no-no. We should be instead adding the plugins as _artifacts_ to these configurations for external consumptions. @breskeby that sound right?

**Review by mark-vieira — CHANGES_REQUESTED** [review]
Left some minor feedback.

File: distribution/docker/build.gradle
Comment (pugnascotia) [reply distribution/docker/build.gradle:312]: As far as I can tell, the release-manager build file just contains what image to pick up, it doesn't do the creation or tagging.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DockerBase.java
Comment (pugnascotia) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DockerBase.java:27]: We reference `base.image` in a number of places in the docker `build.gradle`.

File: plugins/build.gradle
Comment (pugnascotia) [reply plugins/build.gradle]: I swapped this our for artifacts - any better?

File: distribution/docker/build.gradle
Comment (pugnascotia) [reply distribution/docker/build.gradle]: Changed to use a provider.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DockerBase.java
Comment (mark-vieira) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DockerBase.java:27]: Right, but we end up overwriting it [here](https://github.com/elastic/elasticsearch/pull/74980/files#diff-39ead019dc289b9b901bfa7df8efc01b1f040df71c88919b5fc7a6286b0d6597R382). Anyway to avoid that?

File: plugins/build.gradle
Comment (mark-vieira) [reply plugins/build.gradle]: Yup, that should do the trick.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DockerBase.java
Comment (pugnascotia) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DockerBase.java:27]: Oh sorry, you mean in this specific case. Yes, we might as well not set it and comment why.

File: distribution/docker/src/docker/Dockerfile
Comment (pugnascotia) [reply distribution/docker/src/docker/Dockerfile:266]: If it's tricky to modify `allocator.sh` for some reason, we could add logic to `plugin-wrapper.sh` here.

File: distribution/docker/src/docker/Dockerfile
Comment (pugnascotia) [inline distribution/docker/src/docker/Dockerfile:276]: @alpar-t @mieciu I'd still like to know what Cloud actually needs from the Beats distributions. I assume the entire distro isn't required. This isn't essential now, but anything we can trim from the image will help to keep the size down.

**Review by mark-vieira — APPROVED** [review]
:+1:

**Pull request merged**
